working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1374,173 +1580,172 @@ The static type of a pattern `p` being matched against a value of type `M` is:
-            fields, `M` is not `dynamic`, and `M` is not a record with the same
-            number of positional fields.
+*   **Extractor**:
 
-        1.  Calculate the type of each of `f`'s named field subpatterns using
-            the return type of the getter on `M` with the same name as the field
-            as the matched value type. If `M` is `dynamic`, then use `dynamic`
-            as the matched value type. It is a compile-time error if `M` is not
-            `dynamic` and does not have a getter whose name matches the
-            subpattern's field name.
+    1.  Resolve the extractor name to either a type `X` or an enum value of type
+        `X`. It is a compile-time error if the name does not refer to a type or
+        enum value.
 
-            If a named field uses the shorthand syntax to infer a variable
-            subpattern with the same name as the field, then calculate the
-            static type using that inferred variable pattern.
+        **TODO: Can we infer type arguments for `X` from `M`?**
 
-    1.  If `M` is a record type (of any shape) or `dynamic`, then the static
-        type of `p` is a record type whose fields are the fields of `p` with the
-        types of the corresponding subpatterns of `p`. *If the matched value is
-        a record, then we want to ensure that the record pattern has the same
-        shape. We infer a record type from the pattern's fields, then it will be
-        a compile-time error if that inferred record type is not a subtype of
-        the matched value's record type.*
+    1.  Calculate the type of each of `f`'s field subpatterns using the type of
+        the getter on `X` with the same name as the field as the matched value
+        type. It is a compile-time error if `X` does not have a getter whose
+        name matches the subpattern's field name.
 
-    2.  Otherwise, the static type of `p` is `Object?`. *Record patterns are
-        structural and can be applied to values of types that aren't records as
-        long as the type has the right getters.*
+    2.  The static type of `p` is `Object?`. *Extractors exist to check types at
+        runtime, so statically accept all types.*
 
-    3.  It is a compile-time error if `p` has positional fields and `M` is not
-        a record type or `dynamic`. *Only records support positional field
-        destructuring.*
+It is a compile-time error if the type of an expression in a guard clause is not
+`bool` or `dynamic`.
 
-*   **Variable binder**:
+## Refutable and irrefutable patterns
 
-    1.  If the variable has a type annotation, the type of `p` is that type.
+Patterns appear inside a number of other constructs in the language. This
+proposal extends Dart to allow patterns in:
 
-    2.  Else the type of `p` is `M`. *This means that an untyped variable
-        pattern can have its type indirectly inferred from the type of a
-        superpattern:*
+* Top-level and local variable declarations.
```

Comment:
I removed support for top-level variables and static fields when you convinced me to remove support for late variables. Now you've convinced me to drop support for instance fields too.

I guess this will give our future selves something to work on in later releases. :)